### PR TITLE
Add wolfSSL_CTX_set_keylog_callback

### PR DESCRIPTION
**Pull request by TakayukiMatsuo**
Add following APIs for openssl compat layer: 
- wolfSSL_CTX_set_keylog_callback
- wolfSSL_CTX_get_keylog_callback

### CODE REVIEW

File: src/internal.c [hunk 1]
```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+        int mssz   = *secretSz;
+        int hasVal = 0;
+        for (int i = 0; i < mssz; i++) {
+            if (*((byte*)secret) != 0) {
+                hasVal = 1;
+                break;
+            }
+        }
+        if (hasVal == 0)
+            return 0; /* return sd success */
```
Comment (cconlon) [inline src/internal.c:282]: What does "sd" mean here?  Please make comment more clear, thanks.

File: src/internal.c [hunk 2]
```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+
+        /* build up a hex-decoded keylog string
+           "CLIENT_RANDOM <hex-encoded client random> <hex-encoded master-secret>"
+           note that each keylog string does not have LF.
+        */
+        const char* label = "CLIENT_RANDOM";
```
Comment (cconlon) [inline src/internal.c:288]: Variables here should be declared at top of function

File: src/internal.c [hunk 3]
```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+        XMEMSET(log, 0, buffSz);
+        XMEMCPY(log, label, labelSz -1);     /* put label w/o terminator */
+        XMEMSET(log + labelSz - 1, ' ', 1);               /* '\0' -> ' ' */
+        ToHexString(ssl->arrays->clientRandom,
+                        log + labelSz, RAN_LEN);/* put hex encoded random */
+        XMEMSET(log + labelSz + RAN_LEN * 2, ' ', 1);        /* add space*/
+        ToHexString((byte*)secret, log + labelSz + RAN_LEN*2 +1, *secretSz);
+
+        /* pass the log to the client callback*/
+        callback(ssl, (char*)log);
```
Comment (cconlon) [inline src/internal.c:305]: I would prefer "callback" to be slightly more descriptive.  Possibly logCb(), or similar?  Will make it easier to understand when people are reading code.

File: src/internal.c [hunk 4]
```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+    static int SessionSecret_callback_Tls13(WOLFSSL* ssl, int id,
+        const unsigned char* secret, int secretSz, void* ctx)
+    {
+        (void)ctx;
+
+        if (ssl == NULL || secret == NULL || secretSz == 0)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        wolfSSL_CTX_keylog_cb_func callback = ssl->ctx->Keylog_cb;
```
Comment (cconlon) [inline src/internal.c:320]: Same as above, consider changing "callback" variable name.  Possibly "logCb".

File: src/internal.c [hunk 5]
```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+        if (callback == NULL)
+            return 0;
+
+        char  label[50];
```
Comment (cconlon) [inline src/internal.c:324]: Declare variables at top of function, thanks.

File: src/ssl.c [hunk 6]
```diff
@@ -53643,4 +53643,29 @@ int wolfSSL_set_ephemeral_key(WOLFSSL* ssl, int keyAlgo,
 
 #endif /* WOLFSSL_STATIC_EPHEMERAL */
 
+#if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
+/*
+ * This API accepts an user callback which puts key-log records into
```
Comment (cconlon) [inline src/ssl.c:53648]: "an user" should be "a user"

File: src/ssl.c [hunk 7]
```diff
@@ -53643,4 +53643,29 @@ int wolfSSL_set_ephemeral_key(WOLFSSL* ssl, int keyAlgo,
...
+ * a KEY LOGFILE. The callback is stored into a CTX and propagated to
+ * each SSL oject on it's creation timing.
```
Comment (cconlon) [inline src/ssl.c:53650]: "oject" should be "object", "it's" should be "its"

File: src/ssl.c [hunk 8]
```diff
@@ -53643,4 +53643,29 @@ int wolfSSL_set_ephemeral_key(WOLFSSL* ssl, int keyAlgo,
...
+ */
+void wolfSSL_CTX_set_keylog_callback(WOLFSSL_CTX* ctx, wolfSSL_CTX_keylog_cb_func cb)
+{
+    WOLFSSL_ENTER("wolfSSL_CTX_set_keylog_callback");
+  /* stores the callback into WOLFSSL_CTX */
+    if (ctx != NULL) {
+        ctx->Keylog_cb = cb;
```
Comment (cconlon) [inline src/ssl.c:53657]: To match existing naming conventions of callback variables in WOLFSSL_CTX, let's change "Keylog_cb" to "keyLogCb".

File: tests/api.c [hunk 9]
```diff
@@ -28107,6 +28107,359 @@ static void test_wolfSSL_CTX_add_client_CA(void)
...
+
+static void keyLog_callback(const WOLFSSL* ssl, const char* line )
+{
+
+    AssertNotNull(ssl);
+    AssertNotNull(line);
+
+    XFILE fp;
+    const byte  lf = '\n';
+    fp = XFOPEN("./MyKeyLog.txt", "a");
```
Comment (cconlon) [inline tests/api.c:28202]: We should delete this file (MyKeyLog.txt) when users run "make clean".  Please add "MyKeyLog.txt" to CLEANFILES list in Makefile.am

File: wolfssl/internal.h [hunk 10]
```diff
@@ -3002,6 +3002,9 @@ struct WOLFSSL_CTX {
     Srp*  srp;  /* TLS Secure Remote Password Protocol*/
     byte* srp_password;
 #endif
+#if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
+    wolfSSL_CTX_keylog_cb_func Keylog_cb;
```
Comment (cconlon) [inline wolfssl/internal.h:3006]: See note above, let's change "Keylog_cb" name to "keyLogCb" to match callback naming scheme in WOLFSSL_CTX.

**Review by cconlon — CHANGES_REQUESTED** [review]

File: src/internal.c [hunk 11]
```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+    {
+        if (bin == NULL || hex == NULL || BinSz < 2)
+            return;
+
+        byte HexTbl[16] = { '0','1','2','3','4','5','6','7',
+                            '8','9','a','b','c','d','e','f' };
+
+        byte UpNibble, LwNibble;
+
+        for (int i = BinSz; i > 0; i--) {
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply src/internal.c:244]: Moved variables to the top of the function.

File: src/internal.c [hunk 12]
```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+        if (ssl == NULL || secret == NULL || *secretSz == 0)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        wolfSSL_CTX_keylog_cb_func callback = ssl->ctx->Keylog_cb;
+        if (callback == NULL)
+            return 0;
+
+        /* need to make sure the given master-secret has a meaningful value */
+        int mssz   = *secretSz;
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply src/internal.c:273]: Moved variables to the top of the function.

File: src/internal.c (hunk 1 above)
Comment (TakayukiMatsuo) [reply src/internal.c:282]: It was a typo. Changed comment.

File: src/internal.c (hunk 2 above)
Comment (TakayukiMatsuo) [reply src/internal.c:288]: Moved variables to the top of the function.

File: src/internal.c (hunk 3 above)
Comment (TakayukiMatsuo) [reply src/internal.c:305]: Renamed it.

File: src/internal.c (hunk 4 above)
Comment (TakayukiMatsuo) [reply src/internal.c:320]: Renamed it.

File: src/internal.c (hunk 5 above)
Comment (TakayukiMatsuo) [reply src/internal.c:324]: Moved variables to the top of the function.

File: src/ssl.c (hunk 6 above)
Comment (TakayukiMatsuo) [reply src/ssl.c]: Fixed it.

File: src/ssl.c (hunk 7 above)
Comment (TakayukiMatsuo) [reply src/ssl.c]: Fixed them.

File: src/ssl.c (hunk 8 above)
Comment (TakayukiMatsuo) [reply src/ssl.c]: Changed to "keyLogCb".

File: tests/api.c (hunk 9 above)
Comment (TakayukiMatsuo) [reply tests/api.c]: Added the file to Makefile.am.

File: wolfssl/internal.h (hunk 10 above)
Comment (TakayukiMatsuo) [reply wolfssl/internal.h]: Changed to "keyLogCb".

File: src/internal.c [hunk 13]
```diff
@@ -222,6 +222,191 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+
+    static void ToHexString(const byte* bin, byte* hex, int BinSz);
+    static int  SessionSecret_callback(WOLFSSL* ssl, void* secret,
+                                                  int* secretSz, void* ctx);
+    static int  SessionSecret_callback_Tls13(WOLFSSL* ssl, int id,
+                       const unsigned char* secret, int secretSz, void* ctx);
+
+    /* Convert input byte array to Hex string then add a null-terminator
+     *
+     * The buffer for hex must be equal or grater than (bin size * 2 + 1).
```
Comment (cconlon) [inline src/internal.c:235]: Spelling - "grater" should be "greater"

File: tests/api.c [hunk 14]
```diff
@@ -28238,6 +28238,359 @@ static void test_wolfSSL_CTX_add_client_CA(void)
...
+    SSL_CTX_free(ctx);
+    printf(resultFmt, passed);
+
+#endif /* OPENSSL_EXTRA && HAVE_SECRET_CALLBACK */
+}
+static void test_wolfSSL_Tls12_Key_Logging_test(void)
+{
+ #if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
+ /* This test is intended for checking whether keylog callback is called
+  * in client during TLS handshake between tje client and a server.
```
Comment (cconlon) [inline tests/api.c:28375]: Spelling - "tje" should be "the"

File: tests/api.c [hunk 15]
```diff
@@ -28238,6 +28238,359 @@ static void test_wolfSSL_CTX_add_client_CA(void)
...
+    /* check if the keylog file exists */
+
+    char  buff[300] = {0};
+    int  found = 0;
+
+    fp = XFOPEN("./MyKeyLog.txt", "r");
+
+    AssertNotNull(fp);
+
+    while(XFGETS( buff, 300,fp) != NULL ) {
```
Comment (cconlon) [inline tests/api.c:28462]: Let's replace 300 with "(int)sizeof(buff)" here

File: tests/api.c [hunk 16]
```diff
@@ -28238,6 +28238,359 @@ static void test_wolfSSL_CTX_add_client_CA(void)
...
+        if(0 == strncmp(buff,"CLIENT_RANDOM ",
+                    sizeof("CLIENT_RANDOM ")-1)) {
+            found = 1;
+            break;
+        }
+    }
+    XFCLOSE(fp);
+    /* a log starting with "CLIENT_RANODOM " should exit in the file */
```
Comment (cconlon) [inline tests/api.c:28470]: Spelling - "CLIENT_RANODOM" should be "CLIENT_RANDOM"

File: tests/api.c [hunk 17]
```diff
@@ -28238,6 +28238,359 @@ static void test_wolfSSL_CTX_add_client_CA(void)
...
+    AssertNotNull( found );
+    printf(resultFmt, passed);
+
+#endif /* OPENSSL_EXTRA && HAVE_SECRET_CALLBACK */
+}
+static void test_wolfSSL_Tls13_Key_Logging_test(void)
+{
+ #if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
+ /* This test is intended for checking whether keylog callback is called
+  * in client during TLS handshake between tje client and a server.
```
Comment (cconlon) [inline tests/api.c:28480]: Spelling - "tje" should be "the"

File: tests/api.c [hunk 18]
```diff
@@ -28238,6 +28238,359 @@ static void test_wolfSSL_CTX_add_client_CA(void)
...
+    /* check if the keylog file exists */
+
+    char  buff[300] = {0};
+    int  found[4]   = {0};
+
+    fp = XFOPEN("./MyKeyLog.txt", "r");
+
+    AssertNotNull(fp);
+
+    while(XFGETS( buff, 300,fp) != NULL ) {
```
Comment (cconlon) [inline tests/api.c:28564]: Let's replace 300 with "(int)sizeof(buff)"

**Review by cconlon — CHANGES_REQUESTED** [review]
Thanks, just a few more minor requests

File: tests/api.c (hunk 18 above)
Comment (TakayukiMatsuo) [reply tests/api.c:28564]: Fixed it.

File: tests/api.c (hunk 17 above)
Comment (TakayukiMatsuo) [reply tests/api.c:28480]: Fixed it.

File: tests/api.c (hunk 16 above)
Comment (TakayukiMatsuo) [reply tests/api.c:28470]: Fixed it.

File: tests/api.c (hunk 15 above)
Comment (TakayukiMatsuo) [reply tests/api.c:28462]: Fixed it.

File: tests/api.c (hunk 14 above)
Comment (TakayukiMatsuo) [reply tests/api.c:28375]: Fixed it.

File: src/internal.c (hunk 13 above)
Comment (TakayukiMatsuo) [reply src/internal.c:235]: Fixed it.

File: src/internal.c [hunk 19]
```diff
@@ -222,6 +222,191 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+                                                  int* secretSz, void* ctx);
+    static int  SessionSecret_callback_Tls13(WOLFSSL* ssl, int id,
+                       const unsigned char* secret, int secretSz, void* ctx);
+
+    /* Convert input byte array to Hex string then add a null-terminator
+     *
+     * The buffer for hex must be equal or greater than (bin size * 2 + 1).
+     *
+     */
+    static void ToHexString(const byte* bin, byte* hex, int BinSz)
```
Comment (cconlon) [inline src/internal.c:238]: As @JacobBarthelmeh pointed out in PR #3794, can we use Base16_Encode() here in place of ToHexString()?

**Review by cconlon — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 19 above)
Comment (TakayukiMatsuo) [reply src/internal.c:238]: Replaced ToHexString with Base16_Encode.

File: src/internal.c [hunk 20]
```diff
@@ -222,6 +225,178 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+
+        if (log == NULL)
+            return MEMORY_E;
+
+        XMEMSET(log, 0, buffSz);
+        XMEMCPY(log, label, labelSz -1);     /* put label w/o terminator */
+        XMEMSET(log + labelSz - 1, ' ', 1);               /* '\0' -> ' ' */
+
+        outSz = buffSz - labelSz;
+        Base16_Encode(ssl->arrays->clientRandom, RAN_LEN,
```
Comment (cconlon) [inline src/internal.c:292]: Need to check return code of Base16_Encode()

File: src/internal.c [hunk 21]
```diff
@@ -222,6 +225,178 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+                                            log + labelSz, &outSz);
+        XMEMSET(log + labelSz + RAN_LEN * 2, ' ', 1);  /* add space*/
+        outSz = buffSz - labelSz - RAN_LEN * 2 -1;
+        Base16_Encode((byte*)secret, *secretSz,
```
Comment (cconlon) [inline src/internal.c:296]: Need to check return code of Base16_Encode()

File: src/internal.c [hunk 22]
```diff
@@ -222,6 +225,178 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+        log    = XMALLOC(buffSz, ssl->heap, DYNAMIC_TYPE_SECRET);
+        if (log == NULL)
+            return MEMORY_E;
+
+        XMEMSET(log, 0, buffSz);
+        XMEMCPY(log, label, labelSz - 1);     /* put label w/o terminator */
+        XMEMSET(log + labelSz - 1, ' ', 1);               /* '\0' -> ' ' */
+
+        outSz = buffSz - labelSz;
+        Base16_Encode(ssl->arrays->clientRandom, RAN_LEN,
```
Comment (cconlon) [inline src/internal.c:385]: Need to check return code of Base16_Encode()

File: src/internal.c [hunk 23]
```diff
@@ -222,6 +225,178 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+                                                    log + labelSz, &outSz);
+        XMEMSET(log + labelSz + RAN_LEN * 2, ' ', 1);        /* add space*/
+        outSz = buffSz - labelSz - RAN_LEN * 2 - 1;
+        Base16_Encode((byte*)secret, secretSz,
```
Comment (cconlon) [inline src/internal.c:389]: Need to check return code of Base16_Encode()

**Review by cconlon — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 23 above)
Comment (TakayukiMatsuo) [reply src/internal.c:389]: Added return code check.

File: src/internal.c (hunk 22 above)
Comment (TakayukiMatsuo) [reply src/internal.c:385]: Added return code check.

File: src/internal.c (hunk 21 above)
Comment (TakayukiMatsuo) [reply src/internal.c:296]: Added return code check.

File: src/internal.c (hunk 20 above)
Comment (TakayukiMatsuo) [reply src/internal.c:292]: Added return code check.

**Review by cconlon — APPROVED** [review]

**Comment by cconlon** [comment]
Assigning to Todd for final review/merge.

File: src/internal.c [hunk 24]
```diff
@@ -6024,6 +6223,16 @@ int InitSSL(WOLFSSL* ssl, WOLFSSL_CTX* ctx, int writeDup)
     ssl->tls13SecretCtx = NULL;
 #endif
 #endif
+#if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
```
Comment (toddouska) [inline src/internal.c:6226]: IsTLS() determines whether the current connection is TLS or not.

Adding callback setters is not correct here.  Do that in an initialize function.

File: src/internal.c [hunk 25]
```diff
@@ -24591,6 +24800,15 @@ int SendClientKeyExchange(WOLFSSL* ssl)
                 }
                 ssl->options.clientState = CLIENT_KEYEXCHANGE_COMPLETE;
             }
+        #if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
+            if (ssl->sessionSecretCb != NULL) {
+                int secretSz = SECRET_LEN;
+                ret = ssl->sessionSecretCb(ssl, ssl->arrays->masterSecret,
```
Comment (toddouska) [inline src/internal.c:24806]: Why is session secret cb used for the key log cb?  How do you know the key log cb is even set up?  What if the user wants both?

Why is the key log cb only called here on client key exchange send?  There are several places where the client and server generate key material.  Where are those handled?

Comment (TakayukiMatsuo) [reply src/internal.c:24806]: The session secret callback doesn't seem to be OpenSSL compatible, but I thought the keylog cb and session secret cb were provided for almost the same purpose. So I thought we should use the session secret Cb function as much as possible. It was confirmed that it works by connecting the key log Cb to the existing session secret Cb. There are two problems with this strategy:
1. As you pointed out, the session secret Cb and key log Cb that are set later have priority. However, customers who want to set both in the same program are likely to be convinced.
2. The call points for the existing session secret Cb are at DoServerHello () internal.c and DoTls13ServerHello () in tls13.c. Unfortunately, no masterSecret was generated at this call timing. So I needed to create a new call point, so I chose this SendClientKeyExchange (). And other materials are handled properly with existing session secret Cb call points.

Is the above the answer to your question?

File: src/internal.c (hunk 24 above)
Comment (TakayukiMatsuo) [reply src/internal.c:6226]: maybe, Github is showing you wrong function name. It is not implemented in IsTLS but InitSSL.

**Comment by cconlon** [comment]
Assigning back to @TakayukiMatsuo to refactor so that the keylog callback and secret callback are independent of each other, for users that want to use both.

**Comment by TakayukiMatsuo** [comment]
Refactored so that the keylog callback and secret callback are independent of each other.

**Comment by TakayukiMatsuo** [comment]
retest this please

**Comment by JacobBarthelmeh** [comment]
Can you squash the history on this please. Would like to avoid having "Merge branch" commits in the pull request to keep it easier to view.

**Comment by TakayukiMatsuo** [comment]
Trying to squash them.

**Comment by TakayukiMatsuo** [comment]
Sorry it took a long time to squash, but I think it done.

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]
Thanks for squashing the commits. I gave it a run and tested compared to OpenSSL, just had some minor feedback.

**Comment by dgarske** [comment]
Note: We have an example for the TLS v1.3 key logging here that might be helpful. Looks like you already got that part implemented:
https://github.com/wolfSSL/wolfssl-examples/blob/master/tls/client-tls13.c#L49
https://github.com/wolfSSL/wolfssl-examples/blob/master/tls/server-tls13.c#L50

File: src/internal.c [hunk 26]
```diff
@@ -224,6 +227,225 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+        msSz   = *secretSz;
+        hasVal = 0;
+        for (i = 0; i < msSz; i++) {
+            if (*((byte*)secret) != 0) {
+                hasVal = 1;
+                break;
+            }
+        }
+        if (hasVal == 0)
+            return 0; /* master-secret looks invalid */
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply src/internal.c:286]: At the time of the first implementation, it was implemented to call this SessionSecret_callback in existing ssl.sessionSecretCb. DoServerHello in internal.c calls ssl.sessionSecretCb, but at this time ssl-> session.masterSecret was not created yet and all were 0. So Therefore, some logic is needed to check and ignore this case. If it make an error, the handshake will result in an error. In the latest implementation, ssl.sessionSecretCb doesn't call this callback, so this check logic may not be needed, but it's harmless and is left there.

File: src/internal.c [hunk 27]
```diff
@@ -224,6 +227,225 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+
+        if (ssl == NULL || secret == NULL || *secretSz == 0)
+            return BAD_FUNC_ARG;
+        if (ssl->arrays == NULL || ssl->arrays->clientRandom == NULL)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        logCb = ssl->ctx->keyLogCb;
+        if (logCb == NULL)
+            return 0;
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply src/internal.c:274]: In the OpenSSL implementation, it seems that even if callback is NULL, it returns success without error.

**Comment by TakayukiMatsuo** [comment]
Thanks @dgarske for the examples.

File: src/internal.c [hunk 28]
```diff
@@ -224,6 +227,225 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
...
+        int buffSz;
+        byte* log = NULL;
+        word32 outSz;
+        int idx;
+        int ret;
+        (void)ctx;
+
+        if (ssl == NULL || secret == NULL || *secretSz == 0)
+            return BAD_FUNC_ARG;
+        if (ssl->arrays == NULL || ssl->arrays->clientRandom == NULL)
```
[the remark this answers is not in the document]
Comment (TakayukiMatsuo) [reply src/internal.c:268]: Removed those unnecessary check logic.

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**